src/PrestaShopBundle/Bridge/Controller/ControllerConfigurationFactory.php

```diff
@@ -0,0 +1,91 @@
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShopBundle\Bridge\Controller;
+
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorage;
+use Tools;
+
+/**
+ * Create configuration for controller migrate horizontally
+ */
+class ControllerConfigurationFactory
+{
+    /**
+     * @var TokenStorage
+     */
+    private $tokenStorage;
+
+    public function __construct(
+        TokenStorage $tokenStorage
+    ) {
+        $this->tokenStorage = $tokenStorage;
+    }
+
+    public function create(array $configuration = []): ControllerConfiguration
+    {
+        $configuratorController = new ControllerConfiguration();
+
+        $resolver = new OptionsResolver();
+        $this->configureOptions($resolver);
+        $configuration = $resolver->resolve($configuration);
+
+        $configuratorController->id = $configuration['id'];
+        $configuratorController->controllerName = $configuration['controllerName'];
+        $configuratorController->controllerNameLegacy = $configuration['controllerNameLegacy'];
+        $configuratorController->positionIdentifier = $configuration['positionIdentifier'];
+        $configuratorController->table = $configuration['table'];
+        $configuratorController->user = $this->getUser();
+        $configuratorController->folderTemplate = Tools::toUnderscoreCase(substr($configuratorController->controllerNameLegacy, 5)) . '/';
+
+        return $configuratorController;
+    }
+
+    private function getUser()
+    {
+        if (null === $token = $this->tokenStorage->getToken()) {
+            return null;
+        }
+
+        if (!\is_object($user = $token->getUser())) {
+            return null;
+        }
+
+        return $user;
+    }
+
+    private function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setRequired([
+            'id',
+            'controllerName',
+            'controllerNameLegacy',
+            'positionIdentifier',
+            'table',
+        ]);
+    }
```

Comment:
Perhaps the Builder pattern would be more flexible than Factory pattern ?
Take a look at Symfony Form design: only the class name is provided to the builder that is charge of all the boring work.
Here a builder could accept an empty configuration of any type (according to current page to migrate), and fill this configuration with basic stuff. A specific builder could set specific stuff.

Reply:
I agree with @jolelievre. I don't think a Builder is necessary. Because we only need to instantiate an instance of an object, we don't have significant logic and no extensive configuration

We also don't have many configuration objects, just one in this case. I think we don't need the flexibility of the Builder pattern.